gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcApiExceptionFactory.java

```diff
@@ -67,8 +87,76 @@ ApiException create(Throwable throwable) {
     }
   }
 
-  private ApiException create(Throwable throwable, Status.Code statusCode) {
+  private ApiException create(Throwable throwable, Status.Code statusCode, Metadata metadata) {
     boolean canRetry = retryableCodes.contains(GrpcStatusCode.grpcCodeToStatusCode(statusCode));
-    return ApiExceptionFactory.createException(throwable, GrpcStatusCode.of(statusCode), canRetry);
+    if (metadata == null) {
+      return ApiExceptionFactory.createException(
+          throwable, GrpcStatusCode.of(statusCode), canRetry);
+    }
+    byte[] bytes = metadata.get(Metadata.Key.of(ERROR_DETAIL_KEY, Metadata.BINARY_BYTE_MARSHALLER));
+    if (bytes == null) {
+      return ApiExceptionFactory.createException(
+          throwable, GrpcStatusCode.of(statusCode), canRetry);
+    }
+
+    ErrorDetails.Builder errorDetailsBuilder = ErrorDetails.builder();
+    try {
+      com.google.rpc.Status status = com.google.rpc.Status.parseFrom(bytes);
+      List<String> detailsStringList = new ArrayList<>();
+      for (Any detail : status.getDetailsList()) {
+        Message unpack = unpack(detail);
+        if (unpack instanceof ErrorInfo) {
+          ErrorInfo errorInfo = (ErrorInfo) unpack;
+          errorDetailsBuilder.setDomain(errorInfo.getDomain());
+          errorDetailsBuilder.setReason(errorInfo.getReason());
+          errorDetailsBuilder.setErrorInfoMetadata(errorInfo.getMetadataMap());
+        }
+        detailsStringList.add("type: " + detail.getTypeUrl() + "\n" + unpack.toString());
+      }
+      errorDetailsBuilder.setDetailsStringList(detailsStringList);
+    } catch (InvalidProtocolBufferException | ClassNotFoundException e) {
+      // probably don't do anything?
+      e.printStackTrace();
+    }
+    return ApiExceptionFactory.createException(
+        throwable, GrpcStatusCode.of(statusCode), canRetry, errorDetailsBuilder.build());
+  }
+
+  // can we rely on type url to get the classname? How likely is it going to change?
+  private Message unpack(Any detail) throws ClassNotFoundException, InvalidProtocolBufferException {
+    List<String> split = Splitter.on("/").splitToList(detail.getTypeUrl());
+    Class<?> errorType = Class.forName("com." + split.get(split.size() - 1));
+    if (Message.class.isAssignableFrom(errorType)) {
+      return detail.unpack((Class<Message>) errorType);
+    } else {
+      return null;
+    }
+  }
+
+  // Safer but may need to come back and update it in case more error details type added
+  private Message unpackWithHardCodedTypes(Any detail) throws InvalidProtocolBufferException {
```

Comment:
I couldn't find any usages of this method.

Reply:
Used for illustration purpose only to compare with other approaches.